### PR TITLE
API: Disallow strings in logical ufuncs

### LINKED ISSUE #20898: BUG: Vague error message from ufunc `np.any`
### Describe the issue:

Noticed with NumPy 1.22 that the error message from `np.any` is a bit vague. However NumPy 1.21 has a clearer error message. Interestingly the error type raised changed from `TypeError` to `ValueError`.

### Reproduce the code example:

```python
In [1]: import numpy as np

In [2]: a = np.array(["baz", "baz", "baz", "bar", "foo"], dtype="<U3")

In [3]: a.any()
```

### Error message:

The NumPy 1.22 error message is:

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [3], in <module>
----> 1 a.any()

File ~/miniconda/envs/np122/lib/python3.10/site-packages/numpy/core/_methods.py:57, in _any(a, axis, dtype, out, keepdims, where)
     54 def _any(a, axis=None, dtype=None, out=None, keepdims=False, *, where=True):
     55     # Parsing keyword arguments is currently fairly slow, so avoid it for now
     56     if where is True:
---> 57         return umr_any(a, axis, dtype, out, keepdims)
     58     return umr_any(a, axis, dtype, out, keepdims, where=where)

ValueError: invalid literal for int() with base 10: 'baz'
```

Compare this to the previous error message from NumPy 1.21, which was:

```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [3], in <module>
----> 1 a.any()

File ~/miniconda/envs/np121/lib/python3.10/site-packages/numpy/core/_methods.py:57, in _any(a, axis, dtype, out, keepdims, where)
     54 def _any(a, axis=None, dtype=None, out=None, keepdims=False, *, where=True):
     55     # Parsing keyword arguments is currently fairly slow, so avoid it for now
     56     if where is True:
---> 57         return umr_any(a, axis, dtype, out, keepdims)
     58     return umr_any(a, axis, dtype, out, keepdims, where=where)

TypeError: cannot perform reduce with flexible type
```

### NumPy/Python version information:

```
1.22.1 3.10.2 | packaged by conda-forge | (main, Jan 14 2022, 08:02:37) [Clang 11.1.0 ]
```

### PR DESCRIPTION
**Pull request by seberg**
This restores pre 1.22 NumPy behaviour for strings passed into logical
ufuncs.  Logical ufuncs should be able to cast inputs to booleans
without modifying their results.
Thus they do this now (in NumPy 1.22+).  The problem is that string to
bool casts are very strange in NumPy currently.

This should be reverted/removed once string to bool casts are well
defined.  But until then, it seems more reasonable to just reject
strings.

Closes gh-20898

### CODE REVIEW

File: numpy/core/tests/test_ufunc.py [hunk 1]
```diff
@@ -2134,7 +2134,7 @@ def test_logical_ufuncs_mixed_object_signatures(self, ufunc, signature):
             [np.logical_and, np.logical_or, np.logical_xor])
     def test_logical_ufuncs_support_anything(self, ufunc):
         # The logical ufuncs support even input that can't be promoted:
-        a = np.array('1')
+        a = np.array(b'1', dtype="V3")
```
Comment (seberg) [inline numpy/core/tests/test_ufunc.py:2137]: Void tests the same thing and defines the bool cast somewhat reasonably (anything that is not all NULL)

File: numpy/core/src/umath/dispatching.c [hunk 2]
```diff
@@ -1018,6 +1018,11 @@ logical_ufunc_promoter(PyUFuncObject *NPY_UNUSED(ufunc),
         /* bail out, this is _only_ to give future/deprecation warning! */
         return -1;
     }
+    if ((op_dtypes[0] != NULL && PyTypeNum_ISSTRING(op_dtypes[0]->type_num))
+            || PyTypeNum_ISSTRING(op_dtypes[1]->type_num)) {
+        /* bail out on strings: currently caststing them to bool is too weird */
```
Comment (rossbar) [inline numpy/core/src/umath/dispatching.c:1023]:
```suggestion
        /* bail out on strings: currently casting them to bool is too weird */
```

**Review by rossbar — APPROVED** [review]
LGTM less a typo - if you're done with this @seberg I'm happy to put it in

**Comment by jakirkham** [comment]
Thanks Sebastian for the PR and Ross for the review! 😄

**Pull request merged**